### PR TITLE
Pre-pull hyperkube in VHD

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What this PR does / why we need it**: This is one more step towards allowing deployments with no outbound access. Pre-pull the Hyperkube image in VHD.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: related to #4156 

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
https://msazure.visualstudio.com/One/_build/results?buildId=18309701&_a=summary&view=results

File: parts/k8s/kubernetesinstalls.sh [hunk 1]
```diff
@@ -176,22 +176,23 @@ function installImg() {
 }
 
 function pullHyperkube() {
-    retrycmd_if_failure 60 1 1200 img pull $HYPERKUBE_URL || exit $ERR_K8S_DOWNLOAD_TIMEOUT
-    img unpack -o "/home/rootfs-${KUBERNETES_VERSION}" $HYPERKUBE_URL
-    path=$(find /home/rootfs-${KUBERNETES_VERSION} -name "hyperkube")
+    path="/home/hyperkube-downloads/${KUBERNETES_VERSION}"
+    mkdir -p "$path"
+    pullContainerImage "docker" ${HYPERKUBE_URL}
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply parts/k8s/kubernetesinstalls.sh]: Good point. It probably does in the edge case where the hyperkube download happens at runtime (ie. the version is not present in VHD) AND docker is not installed

File: parts/k8s/kubernetesinstalls.sh [hunk 2]
```diff
@@ -175,34 +175,46 @@ function installImg() {
...
-    if [[ ! -f "/usr/local/bin/kubelet-${KUBERNETES_VERSION}" ]]; then
-        installImg
-        pullHyperkube
+function installKubeletAndKubectl() {
+    if [[ ! -f "/usr/local/bin/kubectl-${KUBERNETES_VERSION}" ]]; then
+        if [[ "$CONTAINER_RUNTIME" == "docker" ]]; then
+            extractHyperkube "docker"
+        else
+            installImg
+            extractHyperkube "img"
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesinstalls.sh]: @jackfrancis added back img pull when runtime isn't docker

File: parts/k8s/kubernetesinstalls.sh [hunk 3]
```diff
@@ -175,34 +175,46 @@ function installImg() {
     retrycmd_get_executable 120 5 $img_filepath "https://acs-mirror.azureedge.net/img/img-linux-amd64-v0.4.6" ls || exit $ERR_IMG_DOWNLOAD_TIMEOUT
 }
 
-function pullHyperkube() {
-    retrycmd_if_failure 60 1 1200 img pull $HYPERKUBE_URL || exit $ERR_K8S_DOWNLOAD_TIMEOUT
-    img unpack -o "/home/rootfs-${KUBERNETES_VERSION}" $HYPERKUBE_URL
-    path=$(find /home/rootfs-${KUBERNETES_VERSION} -name "hyperkube")
+function extractHyperkube() {
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesinstalls.sh]: This function is the one doing the extraction really so renamed to extractHyperkube

**Comment by CecileRobertMichon** [comment]
https://msazure.visualstudio.com/One/_build/results?buildId=18358081&_a=summary&view=logs

File: parts/k8s/kubernetesinstalls.sh [hunk 4]
```diff
@@ -175,34 +175,46 @@ function installImg() {
...
+    CLI_TOOL=$1
+    path="/home/hyperkube-downloads/${KUBERNETES_VERSION}"
+    mkdir -p "$path"
+    pullContainerImage $CLI_TOOL ${HYPERKUBE_URL}
+    docker run --rm -v $path:$path {{WrapAsVariable "kubernetesHyperkubeSpec"}} /bin/bash -c "cp /hyperkube $path"
```
Comment (jackfrancis) [inline parts/k8s/kubernetesinstalls.sh]: How will this work if `docker` isn't installed?

Comment (CecileRobertMichon) [reply parts/k8s/kubernetesinstalls.sh]: my bad you're right this won't work... Fix coming

Comment (CecileRobertMichon) [reply parts/k8s/kubernetesinstalls.sh]: fixed

**Pull request merged**